src/tactic/positivity.lean

```diff
@@ -297,18 +297,41 @@ private lemma div_nonneg_of_nonneg_of_pos [linear_ordered_field R] {a b : R} (ha
   0 ≤ a / b :=
 div_nonneg ha hb.le
 
+private lemma int_div_self_pos {a : ℤ} (ha : 0 < a) : 0 < a / a :=
+by { rw int.div_self ha.ne', exact zero_lt_one }
+
+private lemma int_div_nonneg_of_pos_of_nonneg {a b : ℤ} (ha : 0 < a) (hb : 0 ≤ b) : 0 ≤ a / b :=
+int.div_nonneg ha.le hb
+
+private lemma int_div_nonneg_of_nonneg_of_pos {a b : ℤ} (ha : 0 ≤ a) (hb : 0 < b) : 0 ≤ a / b :=
+int.div_nonneg ha hb.le
+
+private lemma int_div_nonneg_of_pos_of_pos {a b : ℤ} (ha : 0 < a) (hb : 0 < b) : 0 ≤ a / b :=
+int.div_nonneg ha.le hb.le
+
 /-- Extension for the `positivity` tactic: division is nonnegative if both numerator and denominator
 are nonnegative, and strictly positive if both numerator and denominator are. -/
 @[positivity]
 meta def positivity_div : expr → tactic strictness
-| `(%%a / %%b) := do -- TODO handle eg `int.div_nonneg`
+| `(%%a / %%b) := do
   strictness_a ← core a,
   strictness_b ← core b,
-  match strictness_a, strictness_b with
-  | (positive pa), (positive pb) := positive <$> mk_app ``div_pos [pa, pb]
-  | (positive pa), (nonnegative pb) := nonnegative <$> mk_app ``div_nonneg_of_pos_of_nonneg [pa, pb]
-  | (nonnegative pa), (positive pb) := nonnegative <$> mk_app ``div_nonneg_of_nonneg_of_pos [pa, pb]
-  | (nonnegative pa), (nonnegative pb) := nonnegative <$> mk_app ``div_nonneg [pa, pb]
+  typ ← infer_type a,
+  match typ, strictness_a, strictness_b with
+  | `(ℤ), positive pa, positive pb :=
+      if a = b then -- Only attempts to prove `0 < a / a`, otherwise falls back to `0 ≤ a / b`
```

Comment:
It should not be too hard to do this right, no? Because `0 < a / b` for `b` positive and `a` positive/nonnegative is true iff `a ≥ b`, which this tactic can prove, if we express as `a - b ≥ 0`.

So I'd define a function `strictness.to_nonneg` which maps `positive` to `nonnegative`, and prove a lemma `int.div_pos_of_sub_nonneg` saying `0 < b → 0 ≤ a - b → 0 < a / b`, then we can write something like:
```lean
  (do pab <- core `(%%a - %%b) >>= strictness.to_nonneg, positive <$> mk_app ``int.div_pos_of_sub_nonneg [pa, pb, pab]) <|>
          nonnegative <$> mk_app ``int_div_nonneg_of_pos_of_pos [pa, pb]
```
here.

Reply:
It's not a question of it being hard, but rather of it being against the tactic spirit. The point is not to prove all `0 ≤` and `0 <` goals but rather to do so structurally. Here you are complicating the expression that's being passed around. This one time is fine, but then someone later might try proving `0 ≤ a - b` from `0 < a / b` or something of sort.

Also, this solves less because we have no `positivity` extension for `-`. I'm not sure such an extension could exist. I think the only hope is that the `norm_num` call within the main `positivity` loop solves it.

What do you think?

Reply:
Okay, if it's really supposed to just be structural then I suppose the current approach is fine.